ambari-common/src/main/python/ambari_commons/repo_manager/__init__.py

```diff
@@ -0,0 +1,69 @@
+"""
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+"""
+__all__ = ["ManagerFactory", "GenericManager"]
+
+import threading
+from ambari_commons import OSCheck, OSConst
+
+from .generic_manager import GenericManager
+from .apt_manager import AptManager
+from .yum_manager import YumManager
+from .zypper_manager import ZypperManager
+from .choco_manager import ChocoManager
+
+
+class ManagerFactory(object):
+  __lock = threading.Lock()
+
+  __repo_manager = None
+  """:type GenericManager"""
+
+  @classmethod
+  def get(cls):
+    """
+    Return Repository Manager object for current OS in safe manner.
+
+    :rtype GenericManager
+    """
+    if not cls.__repo_manager:
+      with cls.__lock:
+        if not cls.__repo_manager:
+          cls.__repo_manager = cls.get_new_instance(OSCheck.get_os_family())
```

Comment:
I take it that Python doesn't have problems with double-checked locking ... it probably encourages it :)

Reply:
the problem before with sub-process.popen lock fix  making me curious about this, so it is just to be sure that it is safe, have no idea if i need to remove it